src/libcore/pin.rs

```diff
@@ -94,6 +95,106 @@
+//! in your type (for example by implementing some operation on `Pin<&[mut] Self>`)
+//! has consequences for your `Drop` implemenetation as well.
+//!
+//! # Projections and Structural Pinning
+//!
+//! One interesting question arises when considering pinning and "container types" --
+//! types such as `Vec` or `Box` but also `RefCell`; types that serve as wrappers
+//! around other types.  When can such a type have a "projection" operation, an
+//! operation with type `fn(Pin<&[mut] Container<T>>) -> Pin<&[mut] T>`?
+//! This does not just apply to generic container types, even for normal structs
+//! the question arises whether `fn(Pin<&[mut] Struct>) -> Pin<&[mut] Field>`
+//! is an operation that can be soundly added to the API.
+//!
+//! This question is closely related to the question of whether pinning is "structural":
+//! when you have pinned a container, have you pinned its contents? Adding a
+//! projection to the API answers that question with a "yes" by offering pinned access
+//! to the contents.
+//!
+//! In general, as the author of a type you get to decide whether pinning is structural, and
+//! whether projections are provided. However, there are a couple requirements to be
+//! upheld when adding projection operations:
+//!
+//! 1. The container must only be [`Unpin`] if all its fields are `Unpin`. This is the default,
+//!    but `Unpin` is a safe trait, so as the author of the container it is your responsibility
+//!    *not* to add something like `impl<T> Unpin for Container<T>`. (Notice that adding a
+//!    projection operation requires unsafe code, so the fact that `Unpin` is a safe trait
+//!    does not break the principle that you only have to worry about any of this if
+//!    you use `unsafe`.)
+//! 2. The destructor of the container must not move out of its argument. This is the exact
+//!    point that was raised in the [previous section][drop-impl]: `drop` takes `&mut self`,
+//!    but the container (and hence its fields) might have been pinned before.
+//!    You have to guarantee that you do not move a field inside your `Drop` implementation.
+//! 3. Your container type must *not* be `#[repr(packed)]`. Packed structs have their fields
+//!    moved around when they are dropped to properly align them, which is in conflict with
+//!    claiming that the fields are pinned when your struct is.
+//! 4. You must make sure that you uphold the [`Drop` guarantee][drop-guarantee]:
+//!    you must make sure that, once your container is pinned, the memory containing the
+//!    content is not overwritten or deallocated without calling the content's destructors.
+//!    This can be tricky, as witnessed by `VecDeque`: the destructor of `VecDeque` can fail
+//!    to call `drop` on all elements if one of the destructors panics. This violates the
+//!    `Drop` guarantee, because it can lead to elements being deallocated without
+//!    their destructor being called.
```

Comment:
This makes it sound like `VecDeque` itself is simply unsound in the presence of `Pin`. Is the intent that `VecDeque` has to do some extra work in this case, or that a past implementation was unsound, or that if your container is implemented with a `VecDeque` you have to do some sort of extra work, or something else? (what extra work could you do to make this sound though? refuse to actually deallocate anything if a destructor threw?)

Reply:
This is in the discussion of what to take care of when adding projections. `VecDeque` has no projections, so it is fine.